test/extensions/quic_listeners/quiche/envoy_quic_alarm_test.cc

```diff
@@ -0,0 +1,148 @@
+  alarm->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  // Advance 9us, alarm shouldn't fire.
+  advanceUsAndLoop(9);
+  EXPECT_FALSE(unowned_delegate->fired());
+  // Advance 1us, alarm should have fired.
+  advanceUsAndLoop(1);
+  EXPECT_TRUE(unowned_delegate->fired());
+  delete alarm;
+
+  unowned_delegate = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm_ptr = alarm_factory_.CreateAlarm(
+      quic::QuicArenaScopedPtr<quic::QuicAlarm::Delegate>(unowned_delegate), &arena_);
+  EXPECT_FALSE(alarm_ptr->IsSet());
+  unowned_delegate = new TestDelegate();
+  alarm_ptr = alarm_factory_.CreateAlarm(
+      quic::QuicArenaScopedPtr<quic::QuicAlarm::Delegate>(unowned_delegate), nullptr);
+  EXPECT_FALSE(alarm_ptr->IsSet());
+}
+
+TEST_F(EnvoyQuicAlarmTest, CreateAlarmAndCancel) {
+  auto unowned_delegate1 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm1(alarm_factory_.CreateAlarm(unowned_delegate1));
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm1->IsSet());
+  auto unowned_delegate2 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm2(alarm_factory_.CreateAlarm(unowned_delegate2));
+  alarm2->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm2->IsSet());
+
+  alarm1->Cancel();
+  EXPECT_FALSE(alarm1->IsSet());
+  // Advance 10us, alarm1 shouldn't fire.
+  advanceUsAndLoop(10);
+  EXPECT_TRUE(unowned_delegate2->fired());
+  EXPECT_FALSE(unowned_delegate1->fired());
+}
+
+TEST_F(EnvoyQuicAlarmTest, CreateAlarmAndReset) {
+  auto unowned_delegate1 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm1(alarm_factory_.CreateAlarm(unowned_delegate1));
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  auto unowned_delegate2 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm2(alarm_factory_.CreateAlarm(unowned_delegate2));
+  alarm2->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm2->IsSet());
+
+  // Reset alarm1 to a different deadline.
+  alarm1->Cancel();
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(5));
+  // Advance 9us, alarm1 should have fired but alarm2 shouldn't.
+  advanceUsAndLoop(9);
+  EXPECT_TRUE(unowned_delegate1->fired());
+  EXPECT_FALSE(unowned_delegate2->fired());
+}
+
+TEST_F(EnvoyQuicAlarmTest, CreateAlarmAndUpdate) {
+  auto unowned_delegate1 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm1(alarm_factory_.CreateAlarm(unowned_delegate1));
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  auto unowned_delegate2 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm2(alarm_factory_.CreateAlarm(unowned_delegate2));
+  alarm2->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm2->IsSet());
+
+  // Update alarm1 to an earlier deadline.
```

Comment:
A common case would be updating to a deadline in the future, can you add a test for that?

Reply:
This test is updating alarm to a future deadline, but just updating it to earlier.

Reply:
I think the common case is 

T=0:  Set( deadline = 10 )
T=5:  Update( deadline = 15 )
T=10:  Update(  deadline = 20 )

The Update() repeats many times without firing, then eventually fire once.

Looks like it's not covered?

Reply:
Thanks, looks great!